### PR TITLE
Fix DH_get_nid() so that it does not cache values.

### PR DESCRIPTION
**Pull request by slontis**
DH_set0_pqg() is now responsible for caching the nid, q and length.

DH with or without named safe prime groups now default to using the maximum private key length (BN_num_bits(q) - 1) when generating a DH private key. The code is now shared between fips and non fips mode for DH key generation.

The OSSL_PKEY_PARAM_DH_PRIV_LEN parameter can be used during keygen to override the maximum private key length to be
in the range (2 * strength ... bits(q) - 1). Where the strength depends on the length of p.

Added q = (p - 1) / 2 safe prime BIGNUMS so that the code is data driven (To simplify adding new names).
The BIGNUMS were code generated.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/dh/dh_lib.c [hunk 1]
```diff
@@ -258,8 +261,11 @@ int DH_set0_key(DH *dh, BIGNUM *pub_key, BIGNUM *priv_key)
         dh->pub_key = pub_key;
     }
     if (priv_key != NULL) {
+        int bits = BN_num_bits(priv_key);
+
         BN_clear_free(dh->priv_key);
         dh->priv_key = priv_key;
+        dh->length = bits;
```
Comment (levitte) [inline crypto/dh/dh_lib.c:268]: It strikes me as a bit odd to use a temporary variable here when you don't up in the `if (q != NULL)` body.

File: test/recipes/15-test_gendh.t [hunk 2]
```diff
@@ -80,4 +80,19 @@ ok(run(app([ 'openssl', 'genpkey',
...
+               'group:ffdhe3072', '-pkeyopt', 'priv_len:3072', '-text'])),
+    'genpkey DH with a large private len should fail');
+
+ ok(run(app([ 'openssl', 'genpkey', '-algorithm', 'DH', '-pkeyopt',
+              'group:ffdhe3072', '-pkeyopt', 'priv_len:256', '-text'])),
+    'genpkey DH with a minimum strength private len');
+
+ ok(run(app([ 'openssl', 'genpkey', '-algorithm', 'DH', '-pkeyopt',
+              'group:ffdhe2048', '-pkeyopt', 'priv_len:224', '-text'])),
+    'genpkey 2048 DH with a minimum strength private len');
```
Comment (levitte) [inline test/recipes/15-test_gendh.t:98]: Please add an LF

**Comment by levitte** [comment]
I very much like what I see.  Thank you for doing this!

I wanna see how this all fares through the CIs before approving, but am otherwise ready to do so

File: crypto/bn/bn_dh.c [hunk 3]
```diff
@@ -45,9 +45,23 @@ static const BN_ULONG modp_1536_p[] = {
     BN_DEF(0x8A67CC74, 0x29024E08), BN_DEF(0x80DC1CD1, 0xC4C6628B),
     BN_DEF(0x2168C234, 0xC90FDAA2), BN_DEF(0xFFFFFFFF, 0xFFFFFFFF)
 };
+/* q = (p - 1) / 2 */
```
Comment (paulidale) [inline crypto/bn/bn_dh.c:48]: Blank lines between these blocks of numbers perhaps?

File: crypto/dh/dh_group_params.c [hunk 4]
```diff
@@ -139,9 +126,6 @@ static DH *dh_new_by_group_name(OPENSSL_CTX *libctx, const char *name)
 
     for (i = 0; i < (int)OSSL_NELEM(dh_named_groups); ++i) {
         if (strcasecmp(dh_named_groups[i].name, name) == 0) {
-            int max_target_security_strength =
-                ifc_ffc_compute_security_bits(dh_named_groups[i].nbits);
-
             /*
              * The last parameter specified here is
```
Comment (paulidale) [inline crypto/dh/dh_group_params.c:130]: This comment needs fixing with the last parameter being removed.

File: crypto/dh/dh_group_params.c [hunk 5]
```diff
@@ -207,63 +177,43 @@ int ffc_set_group_pqg(FFC_PARAMS *ffc, const char *group_name)
...
-            /* Verify q is correct if it exists */
-            if (dh->params.q != NULL) {
-                if (BN_cmp(dh->params.q, q) != 0)
-                    break;  /* returns nid = NID_undef if q does not match */
-            } else {
-                /* assign the calculated q */
-                dh->params.q = q;
-                q = NULL; /* set to NULL so it is not freed */
-            }
+        /* Keep searching until a matching p and g is found */
```
Comment (paulidale) [inline crypto/dh/dh_group_params.c:195]: Nice rewrite of the loop.

File: doc/man3/DH_get0_pqg.pod [hunk 6]
```diff
@@ -57,10 +57,7 @@ after this function has been called. The I<q> parameter may be NULL.
 DH_set0_pqg() also checks if the parameters associated with I<p> and I<g> and
 optionally I<q> are associated with known safe prime groups. If it is a safe
 prime group then the value of I<q> will be set to q = (p - 1) / 2 if I<q> is NULL.
-For safe prime groups the optional length parameter I<length> is set to twice
-the value of the maximum_target_security_strength(BN_num_bits(I<p>)) as listed in
-SP800-56Ar3 Table(s) 25 & 26. If it is not a safe prime group then the optional
-length parameter will be set if I<q> is not NULL to BN_num_bits(I<q>).
+The optional length parameter will be set if I<q> is not NULL to BN_num_bits(I<q>).
```
Comment (paulidale) [inline doc/man3/DH_get0_pqg.pod:60]: Ugh.   What about:

`The optional length parameter will be set to BN_num_bits(I<q>) if I<q> is not NULL`

or

`If I<q> is not NULL, the optional length parameter will be set to BN_num_bits(I<q>)`

File: crypto/dh/dh_lib.c (hunk 1 above)
Comment (slontis) [reply crypto/dh/dh_lib.c:268]: Ahh yeah I changed my mind half way through - will fix..

File: include/crypto/bn_dh.h [hunk 7]
```diff
@@ -9,8 +9,8 @@
 
 #define declare_dh_bn(x) \
     extern const BIGNUM _bignum_dh##x##_p;              \
+    extern const BIGNUM _bignum_dh##x##_q;              \
```
Comment (paulidale) [inline include/crypto/bn_dh.h:12]: What wrong with pgq? :smile:


I think that it's better as pqg too.

File: providers/implementations/keymgmt/dh_kmgmt.c [hunk 8]
```diff
@@ -476,6 +477,10 @@ static int dh_gen_set_params(void *genctx, const OSSL_PARAM params[])
         if (gctx->md == NULL)
             return 0;
     }
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_DH_PRIV_LEN);
+    if (p != NULL
```
Comment (paulidale) [inline providers/implementations/keymgmt/dh_kmgmt.c:481]: Does this fit on one line?

Comment (slontis) [reply providers/implementations/keymgmt/dh_kmgmt.c:481]: yes - will change - will aplly it to the rest of the function also if that is preferred

File: crypto/bn/bn_dh.c (hunk 3 above)
Comment (slontis) [reply crypto/bn/bn_dh.c:48]: This is a deliberate grouping.

File: crypto/dh/dh_group_params.c (hunk 4 above)
Comment (slontis) [reply crypto/dh/dh_group_params.c:130]: Ahh yes - it was in my first pass before my machine shutdown and corrupted this file :)

File: doc/man3/DH_get0_pqg.pod (hunk 6 above)
Comment (slontis) [reply doc/man3/DH_get0_pqg.pod:60]: Ok..

File: crypto/bn/bn_dh.c (hunk 3 above)
Comment (paulidale) [reply crypto/bn/bn_dh.c:48]: The reasonable.  However `dh2048_224_g` has a (pre-existing) blank line before it :(  I didn't notice any others.

**Comment by slontis** [comment]
Rebased to fix a minor conflict in a pod file. Approval should still hold.

**Comment by slontis** [comment]
Minor spacing and comment update for bn_dh.c (Also a trivial change).

**Comment by slontis** [comment]
Travis error is unrelated - caused by another merge.

**Comment by slontis** [comment]
Merged to master. Thanks.

**Pull request closed without merging**